File: lib/models/booking.dart
class Booking {
  final String id;
  final String userId;
  final String itemId;
  final String itemType;
  final String sellerId;
  final DateTime startDate;
  final DateTime endDate;
  final double pricePerUnit;
  final int duration;
  final String durationType;
  final double totalPrice;
  final String status;
  final String? notes;
  final String? rejectionReason;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Booking({
    required this.id,
    required this.userId,
    required this.itemId,
    required this.itemType,
    required this.sellerId,
    required this.startDate,
    required this.endDate,
    required this.pricePerUnit,
    required this.duration,
    required this.durationType,
    required this.totalPrice,
    this.status = 'pending',
    this.notes,
    this.rejectionReason,
    required this.createdAt,
    this.updatedAt,
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['_id'] ?? json['id'] ?? '',
      userId: json['userId'] ?? '',
      itemId: json['itemId'] ?? '',
      itemType: json['itemType'] ?? '',
      sellerId: json['sellerId'] ?? '',
      startDate: json['startDate'] != null
          ? DateTime.parse(json['startDate'])
          : DateTime.now(),
      endDate: json['endDate'] != null
          ? DateTime.parse(json['endDate'])
          : DateTime.now(),
      pricePerUnit: (json['pricePerUnit'] ?? 0).toDouble(),
      duration: json['duration'] ?? 1,
      durationType: json['durationType'] ?? 'hour',
      totalPrice: (json['totalPrice'] ?? 0).toDouble(),
      status: json['status'] ?? 'pending',
      notes: json['notes'],
      rejectionReason: json['rejectionReason'],
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      updatedAt:
          json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'userId': userId,
      'itemId': itemId,
      'itemType': itemType,
      'sellerId': sellerId,
      'startDate': startDate.toIso8601String(),
      'endDate': endDate.toIso8601String(),
      'pricePerUnit': pricePerUnit,
      'duration': duration,
      'durationType': durationType,
      'totalPrice': totalPrice,
      'status': status,
      'notes': notes,
      'rejectionReason': rejectionReason,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }

  // Helper getters
  bool get isPending => status == 'pending';
  bool get isConfirmed => status == 'confirmed';
  bool get isCompleted => status == 'completed';
  bool get isCancelled => status == 'cancelled';

  bool get isActive => isConfirmed && endDate.isAfter(DateTime.now());
  bool get isPast => endDate.isBefore(DateTime.now());

  String get statusText {
    switch (status) {
      case 'pending':
        return 'Күтуде';
      case 'confirmed':
        return 'Расталған';
      case 'completed':
        return 'Аяқталған';
      case 'cancelled':
        return 'Болдырылмаған';
      default:
        return status;
    }
  }

  String get itemTypeText {
    switch (itemType) {
      case 'instrument':
        return 'Аспап';
      case 'stage':
        return 'Сахна';
      case 'studio':
        return 'Студия';
      default:
        return itemType;
    }
  }
}

File: lib/server/models/Booking.js
const mongoose = require('mongoose');

const bookingSchema = new mongoose.Schema({
  // User who is renting
  userId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },

  // Item being rented
  itemId: {
    type: mongoose.Schema.Types.ObjectId,
    required: true,
    index: true
  },

  itemType: {
    type: String,
    enum: ['instrument', 'stage', 'studio'],
    required: true
  },

  // Seller/Owner of the item
  sellerId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true,
    index: true
  },

  // Booking dates
  startDate: {
    type: Date,
    required: true
  },

  endDate: {
    type: Date,
    required: true
  },

  // Pricing details
  pricePerUnit: {
    type: Number,
    required: true,
    min: 0
  },

  duration: {
    type: Number,
    required: true,
    min: 1
  },

  durationType: {
    type: String,
    enum: ['hour', 'day'],
    required: true
  },

  totalPrice: {
    type: Number,
    required: true,
    min: 0
  },

  // Status tracking
  status: {
    type: String,
    enum: ['pending', 'confirmed', 'completed', 'cancelled'],
    default: 'pending'
  },

  // Optional notes from renter
  notes: {
    type: String,
    trim: true,
    maxlength: 500
  },

  // Rejection reason from seller
  rejectionReason: {
    type: String,
    trim: true,
    maxlength: 500
  },

  // Timestamps
  createdAt: {
    type: Date,
    default: Date.now
  },

  updatedAt: {
    type: Date,
    default: Date.now
  }
});

// Indexes for efficient queries
bookingSchema.index({ userId: 1, createdAt: -1 });
bookingSchema.index({ sellerId: 1, createdAt: -1 });
bookingSchema.index({ itemId: 1, itemType: 1 });
bookingSchema.index({ status: 1 });
bookingSchema.index({ startDate: 1, endDate: 1 });

// Compound index for availability checking
bookingSchema.index({ itemId: 1, itemType: 1, startDate: 1, endDate: 1, status: 1 });

// Update timestamp on save
bookingSchema.pre('save', function(next) {
  this.updatedAt = Date.now();
  next();
});

// Virtual for checking if booking is active
bookingSchema.virtual('isActive').get(function() {
  return this.status === 'confirmed' && this.endDate > new Date();
});

// Virtual for checking if booking is past
bookingSchema.virtual('isPast').get(function() {
  return this.endDate < new Date();
});

module.exports = mongoose.model('Booking', bookingSchema);
